### PR TITLE
Drop all capabilities from control-plane containers

### PR DESCRIPTION
**Pull request by michaelbannister**
As suggested by @rlenglet during discussion of #20311. These containers don't run as root
and don't need any capabilities enabled.

### CODE REVIEW

**Review by ayj — APPROVED** [review]

**Comment by michaelbannister** [comment]
failed to create containerd task: OCI runtime create failed: container_linux.go:346: starting container process caused \"chdir to cwd (\\\"/home/nonroot\\\") set in config.json failed: permission denied\": unknown

**Comment by michaelbannister** [comment]
`istio-testing/citadel:istio-testing-distroless` failed starting:
```
failed to create containerd task: OCI runtime create failed: container_linux.go:346: starting container process caused \"chdir to cwd (\\\"/home/nonroot\\\") set in config.json failed: permission denied\": unknown
```

**Review by rlenglet — APPROVED** [review]
Nice, thanks!

**Comment by rlenglet** [comment]
```
2020-02-04T21:45:32.188040Z	info	ci	Checking pods ready...
2020-02-04T21:45:32.192804Z	info	ci	  [ 0]                istio-citadel-7d97bbcf88-krkb7         Running (container not ready: 'citadel')
```

**Comment by michaelbannister** [comment]
The distroless base image has this:
```
Permission     UID:GID       Size  Filetree
drwxr-xr-x         0:0     1.8 MB  └── .
drwxr-xr-x         0:0        0 B      ├── bin
drwxr-xr-x         0:0        0 B      ├── boot
drwxr-xr-x         0:0        0 B      ├── dev
drwxr-xr-x         0:0     259 kB      ├─⊕ etc
drwxr-xr-x 65532:65532        0 B      ├── home
drwx------ 65532:65532        0 B      │   └── nonroot
```

…and trying to run just this base image as user:group 1337:1337 gives the same error:
```
$ docker run --rm -it -u 1337:1337 -w /home/nonroot gcr.io/distroless/static:nonroot /bin/sh
docker: Error response from daemon: OCI runtime create failed: container_linux.go:346: starting container process caused "chdir to cwd (\"/home/nonroot\") set in config.json failed: permission denied": unknown.
```

**Comment by michaelbannister** [comment]
But I don't understand why this is failing now when it didn't for my previous PR…

**Comment by rlenglet** [comment]
Interesting. Why are we relying on the home directory of that user...

**Comment by howardjohn** [comment]
We need to run in `/`, we rely on it to support VMs/development otherwise it tries to write to things we don't have access to

**Comment by michaelbannister** [comment]
> Interesting. Why are we relying on the home directory of that user...

If you look at the [distroless image base](https://github.com/GoogleContainerTools/distroless/blob/master/base/base.bzl#L77) it's the equivalent of setting `WORKDIR /home/nonroot` and that directory has those permissions I showed above. When we specify a UID/GID to the container runtime it seems to try and change directory to /home/nonroot and fails because UID 1337 doesn't have permission?

But AFAICT this hasn't changed recently (including in the base image?)

**Comment by howardjohn** [comment]
Do we just need something like https://github.com/istio/istio/blob/0996b80fd41dc9ac92e6d2129a78a83c68aff76a/pilot/docker/Dockerfile.proxyv2#L46?

**Comment by rlenglet** [comment]
> Do we just need something like https://github.com/istio/istio/blob/0996b80fd41dc9ac92e6d2129a78a83c68aff76a/pilot/docker/Dockerfile.proxyv2#L46?

+1 The `WORKDIR` may be inherited from the base image.

**Comment by michaelbannister** [comment]
I could certainly change the WORKDIR – but what I don't understand is why this failure should show up on this change, where we're just dropping capabilities, rather than my previous change which made the containers run as 1337:1337 ?

**Comment by rlenglet** [comment]
Here are Docker's default capabilities that we're dropping here:
https://github.com/moby/moby/blob/46ec8731fbce9b13e41bd85d6cea691406bba852/oci/caps/defaults.go#L6-L19
```
		"CAP_CHOWN",
		"CAP_DAC_OVERRIDE",
		"CAP_FSETID",
		"CAP_FOWNER",
		"CAP_MKNOD",
		"CAP_NET_RAW",
		"CAP_SETGID",
		"CAP_SETUID",
		"CAP_SETFCAP",
		"CAP_SETPCAP",
		"CAP_NET_BIND_SERVICE",
		"CAP_SYS_CHROOT",
		"CAP_KILL",
		"CAP_AUDIT_WRITE",
```
Maybe we're changing file ownerships somewhere in the code?

**Comment by michaelbannister** [comment]
I think it's not getting as far as running the entrypoint because the UID/GID doesn't allow to change to the workdir.

I built this very minimal image:
```
$ docker build -q -t test-workdir-permissions - <<EOF
FROM tianon/true as source
FROM gcr.io/distroless/static:nonroot
COPY --from=source /true /true
ENTRYPOINT ["/true"]
EOF
```
and then ran it:
```
$ docker run --rm -it -u 1337:1337 test-workdir-permissions
docker: Error response from daemon: OCI runtime create failed: container_linux.go:346: starting container process caused "chdir to cwd (\"/home/nonroot\") set in config.json failed: permission denied": unknown.
$ docker run --rm -it -u 1337:1337 -w / test-workdir-permissions
[works just fine]
```

So – yes, we could change the workdir in the dockerfile. But I still don't get why this was working _before_ my change, and not working once we explicitly drop capabilities?

**Comment by michaelbannister** [comment]
I won't have a chance to do any more on this until Sunday at earliest. If anyone else is keen to pick it up and submit a replacement PR, I shan't be offended!

**Comment by michaelbannister** [comment]
OK, so this appears to be a difference in behaviour between Kind and Kubernetes proper.
Details here for a simple reproducible demonstration:
https://github.com/michaelbannister/distroless-permissions-test

This could mean that the changes in #20311 mean that deployment of the _distroless_ citadel (and maybe other images?) won't actually work in a full Kubernetes cluster? :(

**Comment by michaelbannister** [comment]
This now fixes the problem with the `-distroless` images introduced in #20311 but only discovered here.
This should either make this a candidate for cherrypick into release-1.5, or else #20820 should probably be reverted from release-1.5.

File: pilot/docker/Dockerfile.pilot
Comment (sdake) [inline pilot/docker/Dockerfile.pilot]: redundant `WORKDIR /` results in a small ineffeciency.

Comment (michaelbannister) [reply pilot/docker/Dockerfile.pilot]: It's not redundant at all: it means this image will actually run.

My previous PR (#20311) changed the Deployments so that the containers run as user/group 1337:1337.
Kubernetes (or Docker on its own) can't successfully run a container from this image when setting user/group to 1337:1337 – see [this comment](#issuecomment-582148470) – because the distroless base image has set `WORKDIR /home/nonroot` and user 1337 has no permissions to reach that directory. By changing the workdir to `/` we fix this problem.

Really, this fix should have gone in #20311 but I didn't discover this problem then due to a difference in behaviour with kind compared to Kubernetes. Kind only showed up the same error once I'd configured the container's security context to drop all capabilities (i.e. this PR). I'm planning to raise an issue on the kind project shortly.

Comment (rlenglet) [reply pilot/docker/Dockerfile.pilot]: I think @sdake's point is that we're already doing this 2 lines below in this particular `Dockerfile`. It's still required in the other `Dockerfile`s, I don't think there's any question about that.

Comment (michaelbannister) [reply pilot/docker/Dockerfile.pilot]: Ah, my apologies @sdake – the minimal context shown in the PR view meant I didn't see this! Will remove this line & push.

**Review by howardjohn — APPROVED** [review]

File: pilot/docker/Dockerfile.pilot
Comment (sdake) [reply pilot/docker/Dockerfile.pilot]: @michaelbannister - thanks. I didn't realize the context wasn't shown or I would have offered more information in the review. Using GitHub is a bit robotic for me 😁

**Comment by BenTheElder** [comment]
> OK, so this appears to be a difference in behaviour between Kind and Kubernetes proper.

Nitpicking but, KIND _is_ Kubernetes proper, we ship fully upstream unmodified kubernetes :+)

This appears to be a difference in containerd and dockerd with `WORKDIR` (?), I've looped in a maintainer there to look further after a bit of debug myself,.

We can reproduce the same behavior on GKE with containerd versus dockerd nodes with the internal `chdir` failing on docker or containerd+the cap drop all but not containerd without the cap drop.

Setting the WORKDIR explicitly seems like a sane approach in the meantime imho.

**Comment by BenTheElder** [comment]
Upstream fix pending in https://github.com/containerd/cri/pull/1397, we'll pull it in once it's merged.

**Comment by michaelbannister** [comment]
Still awaiting review approval from @costinm @linsun @rshriram (I'm unclear whether this is OR or AND). This also needs cherry-picking to fix a bug in release 1.5, which can't happen until this is merged to master.

**Comment by howardjohn** [comment]
you need to look at the bottom of the pr not the reviews list - "Waiting on code owner review from istio/wg-policies-and-telemetry-maintainers and/or istio/wg-security-maintainers". You still have the ambiguous and/or though... in this case I happen to knows it's and (based on CODEOWNERS file).

**Review by yangminzhu — APPROVED** [review]

**Review by myidpt — APPROVED** [review]

**Review by mandarjog — APPROVED** [review]
/lgtm

**Comment by michaelbannister** [comment]
I'm preparing a separate PR to fix the cherry-pick into release-1.5. Will link here when it's ready.

**Comment by michaelbannister** [comment]
Note: the merge conflicts were in the generated-but-committed files under `operator/cmd/mesh/testdata/manifest-generate/output`.

**Pull request merged**
